pkg/operator/csidriveroperator/olmremovalcontroller.go

```diff
@@ -0,0 +1,463 @@
+		Reason:  "Finished",
+		Status:  operatorapi.ConditionTrue,
+		Message: message,
+	}
+
+	// Clear the old namespace annotation - the driver has been fully removed.
+	if err := c.saveOperatorNamespace(""); err != nil {
+		return err
+	}
+	_, _, err := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(progressing),
+		v1helpers.UpdateConditionFn(available),
+	)
+	return err
+}
+
+func (c *OLMOperatorRemovalController) ensureOperatorDeploymentRemoved(namespace, name string) (bool, error) {
+	// Do not actively remove the Deployment here, that's OLM's job.
+	_, err := c.deploymentLister.Deployments(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+	klog.V(4).Infof("Deleted Deployment %s/%s", namespace, name)
+	// Don't report the Deployment is removed, wait until IsNotFound error above
+	return false, nil
+}
+
+func (c *OLMOperatorRemovalController) ensureDriverDeploymentRemoved(ctx context.Context, namespace, name string) (bool, error) {
+	_, err := c.deploymentLister.Deployments(namespace).Get(name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err
+		}
+		return true, nil
+	}
+	err = c.kubeClient.AppsV1().Deployments(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+	if err != nil {
+		return false, err
+	}
+	klog.V(4).Infof("Deleted driver Deployment %s/%s", namespace, name)
+	// Don't report the Deployment is removed, wait until IsNotFound error above
+	return false, nil
+}
+
+func (c *OLMOperatorRemovalController) ensureDriverDaemonSetRemoved(ctx context.Context, namespace, name string) (bool, error) {
+	_, err := c.daemonSetLister.DaemonSets(namespace).Get(name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err
+		}
+		return true, nil
+	}
+	err = c.kubeClient.AppsV1().DaemonSets(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+	if err != nil {
+		return false, err
+	}
+	klog.V(4).Infof("Deleted driver DaemonSet %s/%s", namespace, name)
+	// Don't report the DaemonSet is removed, wait until IsNotFound error above
+	return false, nil
+}
+
+func (c *OLMOperatorRemovalController) ensureCRRemoved(ctx context.Context, res schema.GroupVersionResource) (bool, error) {
+	cr, err := c.dynamicClient.Resource(res).Get(ctx, oldCRName, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err
+		}
+		return true, nil
+	}
+	if len(cr.GetFinalizers()) > 0 {
+		cr.SetFinalizers([]string{})
+		cr, err = c.dynamicClient.Resource(res).Update(ctx, cr, metav1.UpdateOptions{})
+		if err != nil {
+			return false, err
+		}
+		klog.V(4).Infof("Deleted old CR finalizers")
+	}
+	err = c.dynamicClient.Resource(res).Delete(ctx, oldCRName, metav1.DeleteOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+	klog.V(4).Infof("Deleted old CR")
+	// Don't report the CR is removed, wait until IsNotFound error above
+	return false, nil
```

Comment:
If `err == nil` can't we assume the CR was deleted correctly? This would speed things up, as we wouldn't have to wait for a resync

Reply:
I'd rather have it confirmed from the API server...